table/block_based/block_based_table_builder.cc

```diff
@@ -904,9 +912,11 @@ void BlockBasedTableBuilder::Add(const Slice& key, const Slice& value) {
       r->first_key_in_next_block = &key;
       Flush();
 
-      if (r->state == Rep::State::kBuffered && r->buffer_limit != 0 &&
-          r->data_begin_offset > r->buffer_limit) {
-        EnterUnbuffered();
+      if (r->state == Rep::State::kBuffered) {
+        if ((r->buffer_limit != 0 && r->data_begin_offset > r->buffer_limit) ||
+            r->buffer_exceeds_global_memory_limit) {
```

Comment:
Question 2:
I decided to do the checking and switching to unbuffer here instead of in WriteBlock() right after we add the data block into the buffer. My motivation is to (1)keep WriteBlock() simple and Flush() easy to resonate (2)align with what "checking against user-set buffer_limit" does since they all follow unbuffering-policy-by-size.

Downside though, is that we need to keep a shared bool variable "buffer_exceeds_global_memory_limit" around and we need to remember to manipulate that. 

Let me know if you see we should do the checking and switching somewhere else.